app/services/analytics.rb

```diff
@@ -71,6 +72,29 @@ def request_attributes
     attributes.merge!(browser_attributes)
   end
 
+  def ab_test_attributes
```

Comment:
From my experience, analytics is one of the bigger pain points with A/B tests, so I like the idea of something that makes setting it up a little easier. I haven't checked the code, but I'd at least assume only users who have actually entered a flow with a A/B test would have a value included here. Maybe we could also turn this into an allowlist of event names corresponding to a specific test?

Reply:
Yeah, this has been a pain point for Team Ada. We have a lot of events, and ensuring that all of them are properly augmented with A/B test data has been a big pain. Not to mention if you miss one or two that you later learn would be useful.

The way this works is:

* The A/B test is configured to return a discriminator value based on request/session/user/etc.
* If the discriminator is `nil`, we don't assign the user to a bucket, and don't include that test in any `ab_tests` we log
* If the user is not assigned to any buckets (e.g. `ab_tests` is `{}`), we just don't even bother logging `ab_tests`

Reply:
I find it a bit of a struggle to find the properties in events when searching them, so I like Andrew's suggestion. I do see the difficulty in defining them ahead of time though. Is there a different middle ground that makes it easier to capture the A/B tests where needed without including it on every event?

Reply:
We could include, as part of the test definition, some kind of allow list or proc that can be used to evaluate whether to log that test's bucket information? Something like:

```ruby
MY_AB_TEST = AbTest.new(
  # ...
  should_log: ->(event_name) { event_name == 'IdV: final resolution' }
  # ...
)
```

You could imagine different variations of this (regex, allow list, etc.), but the idea is the same.

Reply:
I think it's good, but I feel like it will end up being all by default unless we are somewhat restrictive in the options available.

Reply:
I will voice support for adding the properties to every analytics event. We have had issues where the A/B test buckets were logged on events but they turned out to be the incorrect events.

For example, one the 10-digit OTP A/B test it would have been helpful to have the bucket logged on all events so we could have had them on the OTP send events in addition to the confirmation events.

The tests should also be temporary so they will add a property to every event but it will only be present for the duration of the test.

Reply:
Maybe in a future world when the Data Warehouse is more widely available and we have proper JOINs, we can have a singular "bucket assignment" event and then join to that, but yes in the meantime without join-less workflow, every event seems necessary

Reply:
Do we set up the expected funnel for the test prior to the implementation? If not, would that help ensure that the relevant events are covered?

Reply:
> Do we set up the expected funnel for the test prior to the implementation?

Yes, establishing a baseline for the test is the first part of setting up the test. Doing so ensures that we are able to measure that a change is actually taking place. However, these results often lead to further questions and we need to go back and add buckets to more events causing the test condition to exist for longer, especially with our long cycle times. 2 examples that come to mind are the single-page welcome / agreement step and, as mentioned above, the 10-digit OTP test.

Reply:
Makes sense, seems like this is probably the best option